large_page/large_page.cc

```diff
@@ -160,112 +169,124 @@ static bool IsTransparentHugePagesEnabled() {
+    if (ret < 0) {                                      \
+      status = (status == oper##_munmap_tmem_failed)    \
+        ? oper##_munmaps_failed                         \
+        : oper##_munmap_nmem_failed;                    \
+    }                                                   \
+    return status;                                      \
   }
 
   ret = madvise(tmem, size, MADV_HUGEPAGE);
-  if (ret < 0) {
-    PrintSystemError(errno);
-    ret = munmap(tmem, size);
-    if (ret < 0) {
-      PrintSystemError(errno);
-    }
-    ret = munmap(nmem, size);
-    if (ret < 0) {
-      PrintSystemError(errno);
-    }
-
-    return ret;
-  }
+  CLEAN_EXIT_CHECK(map_see_errno_madvise_tmem);
 
   memcpy(start, nmem, size);
   ret = mprotect(start, size, PROT_READ | PROT_EXEC);
-  if (ret < 0) {
-    PrintSystemError(errno);
-    ret = munmap(tmem, size);
-    if (ret < 0) {
-      PrintSystemError(errno);
-    }
-    ret = munmap(nmem, size);
-    if (ret < 0) {
-      PrintSystemError(errno);
-    }
-    return ret;
-  }
+  CLEAN_EXIT_CHECK(map_see_errno_mprotect);
+
+#undef CLEAN_EXIT_CHECK
 
   // Release the old/temporary mapped region
   ret = munmap(nmem, size);
   if (ret < 0) {
-    PrintSystemError(errno);
+    status = map_see_errno_munmap_nmem_failed;
   }
 
-  return ret;
+  return status;
 }
 
 // Align the region to to be mapped to 2MB page boundaries.
-static void AlignRegionToPageBoundary(text_region& r) {
-  r.from = reinterpret_cast<char*>(largepage_align_up(reinterpret_cast<int64_t>(r.from)));
-  r.to = reinterpret_cast<char*>(largepage_align_down(reinterpret_cast<int64_t>(r.to)));
-  r.total_largepages = (r.to - r.from) / hps;
+void AlignRegionToPageBoundary(MemRange* r) {
+  r->from = reinterpret_cast<void*>(LargePageAlignUp(
+                      reinterpret_cast<uintptr_t>(r->from)));
+  r->to = reinterpret_cast<void*>(LargePageAlignDown(
+                      reinterpret_cast<uintptr_t>(r->to)));
+}
+
+MapStatus CheckMemRange(const MemRange& r) {
+  if (r.from == nullptr || r.to == nullptr || r.from > r.to) {
+    return map_invalid_region_address;
+  }
+
+  if (reinterpret_cast<size_t>(r.to) -
+      reinterpret_cast<size_t>(r.from) < hps) {
+    return map_region_too_small;
+  }
+
+  return map_ok;
 }
 
 // Align the region to to be mapped to 2MB page boundaries and then move the
 // region to large pages.
-static int AlignMoveRegionToLargePages(text_region& r) {
-  AlignRegionToPageBoundary(r);
+MapStatus AlignMoveRegionToLargePages(MemRange r) {
+  AlignRegionToPageBoundary(&r);
 
-  if (r.total_largepages < 1) {
-    fprintf(stderr, "Hugepages WARNING: region is too small for large pages\n");
-    return -1;
+  MapStatus status = CheckMemRange(r);
+  if (status != map_ok) {
+    return status;
   }
 
-  if (r.from > reinterpret_cast<void*>(&MoveRegionToLargePages))
+  if (r.from > (void*)MoveRegionToLargePages ||
+      r.to <= (void*)MoveRegionToLargePages) {
     return MoveRegionToLargePages(r);
+  }
 
```

Comment:
The original .lpstub was before the text so the from > made sense. Now I would suggest to keep only the <=

Reply:
Fixed in https://github.com/intel/iodlr/pull/17/commits/ff9b0cacac746cf14380e1da17caebc4d0927833